Add `ember lint` command

alexlafroscia opened the pull request:
`ember lint` run `eslint` on the `app` and `tests` directories.

The only issue right now is that the output of the ESLint command is grabbed from the `STDOUT` of the child process and then passed through to the output of the Ember process, so we lose the original colors.  I would love to maintain them, but I'm just not sure how.  I'd love a pointer in the right direction!

Also, I used `Promise` in the new task; I'm guessing that won't be a problem but if I should write that another way, I'm happy to.

Closes #92

alexlafroscia commented:
Instead of running `eslint` through a child process, I did originally implement this by using ESLint's Node API directly.  However, I wasn't sure if that made sense for a first pass, since we would have to come up with an output format ourselves.

alexlafroscia commented:
Not sure about the right way to test this... Is it necessary?

Turbo87 commented:
@alexlafroscia when we run `ember serve` I see ESLint output too and it is colored. how is that handled? could we maybe reuse that code?

alexlafroscia commented:
Hmm, I'll look into that.

alexlafroscia commented:
I was able to get the ESLint output stream to pipe right into the Ember CLI UI 🙌

Need to clean things up a bit but I was able to get all the color like we should!

alexlafroscia commented:
There we go 😄  

<img width="576" alt="screen shot 2016-07-14 at 5 23 46 pm" src="https://cloud.githubusercontent.com/assets/1645881/16859907/ce819b6e-49e7-11e6-8908-63eb21f31b1a.png">

There is one small visual glitch, where the `Running ESLint` remains on screen after the output is printed to the screen.  I technically tell it to stop the spinner, I think it's just too late to stop it from displaying on screen.  I don't think it's that big of a deal either way, though.

Not sure about the best way to test something like this, I'd love some direction on that @rwjblue

alexlafroscia commented on lib/helpers/path-to-eslint.js:
```diff
@@ -0,0 +1,20 @@
+  return asyncExec('npm', ['bin'])
+    .then(function(location) {
+      return join(
+        location.trim(), '../ember-cli-eslint/node_modules/.bin/eslint'
```
Is there a better way to resolve the location of the `eslint` binary that's installed with this package?  I want to make sure that we use the right one so that the same version is used between `ember test` and `ember lint`

rwjblue replied:
In theory, `require.resolve('eslint')` will give you the location of eslint, then you can traverse the structure to find the associated bin file.

```
var eslintPath = path.join(require.resolve('eslint'), '../bin/eslint.js');
```

Maybe ^^ will work?

alexlafroscia replied:
Yeah, I made that comment looking for help before I remembered `require.resolve`.  That's exactly what I'm doing now.

https://github.com/alexlafroscia/ember-cli-eslint/blob/0fcab0ade6480f47abdba5c17fee7e9c16a09e4d/lib/helpers/eslint-command.js#L8

alexlafroscia commented:
Hmm... not sure why the tests are failing on CI.  I tried installing the addon from my branch (instead of linking), reinstalled all the dependencies, and it worked fine.  Then took my local repo and reinstalled all the dependencies, ran `ember try default test` and `ember try ember-release test` and both of those were fine too.

Turbo87 commented:
@alexlafroscia FYI https://github.com/ember-cli/broccoli-lint-eslint/blob/master/lib/index.js#L180 is where `broccoli-lint-eslint` is writing the ESLint output. can/should we reuse some of the code that's already available in there?

alexlafroscia commented:
I'm not sure about the "should" part.  I guess it depends on whether you want ESLint to be run directly, or through Broccoli.  I took the approach of executing ESLint directly, which I feel makes more sense.  We could have `ember lint` just execute `ember test --filter 'ESLint'` but then we'd get the ESLint output _and_ the test runner pass/fails, which I feel is redundant.  Or I suppose we could just run Broccoli in our own process with that ESLint plugin, but I'm not sure how that would extend to `--fix` then.

If you think running Broccoli to run ESLint is a better approach than just running ESLint, I can do that.  Just not sure what we're gaining by doing that, other than the assurance that we resolve ESLint from the right location.

alexlafroscia commented:
After thinking about it a bunch, I think @Turbo87 is right and we should just rely on the linting already provided by the Broccoli plugin rather than invoking `eslint` directly, since it's going to be tough to resolve the ESLint dependency correctly (which is the problem that Travis seems to be having right now).

I might break this back into two PRs, one for `ember lint` and one for `ember lint:test` since, if we're going to interact with ESLint only through `broccoli-lint-eslint`, I need to add the ability to fix files in the tree to the Broccoli plugin and then perform the fixes through that.

alexlafroscia commented:
Alright, I removed the code for `ember lint:fix` and changed the `ember lint` code to rely on the Broccoli plugin under the hood.  Much less code and more reliable than what I was doing.

I'd like to improve this in the future by making the promise returned by the task `reject` if there was an error, so that we get a non-zero exit code in the console (probably useful).  At the moment, I don't think there's any good way to detect that, but once [this PR](https://github.com/ember-cli/broccoli-lint-eslint/pull/44) lands I think I should be able to.

Turbo87 commented:
LGTM :)

@BrianSipple @rwjblue any comments?

alexlafroscia commented:
Now that `throwOnError` should actually work, I'm going to revisit this and see if I can exit with a non-0 error code if there are any issues.

alexlafroscia commented:
On second thought... `throwOnError` is only available on the latest `3.X` release of ESLint but we're currently consuming `2.X` 😒

Turbo87 commented:
@alexlafroscia how is the eslint 2 cli determining what exit code to use? there has to be some way we can reuse that.

alexlafroscia commented:
The issue, as far as I remember, was that the commit that added the ability to determine if there were any errors (and thus, whether or not we need to exit with a non-zero exit code) didn't get merged into the 2.X branch of `broccoli-lint-eslint`.  However I will take a look and see if that's still the case/if there's something else I can do to determine how to exit.

Turbo87 commented:
can you point me to that commit?

BrianSipple commented:
@Turbo87 I've actually been looking into this and I'm just about to PR a change for 2.x. I can highlight everything there.

alexlafroscia commented:
Sorry this got away from me.  Will try to push through it today

alexlafroscia commented:
Alright, so I rebased off of the current `master` and added the non-zero exit code when there's a failure or error, as well as passing in `ui.writeLine` as the log function like the main plugin does.  Should be all set.  Turns out that `throwOnError` isn't really the behavior that we want anyway, since throwing on the first error would prevent all of the errors from being printed.

Turbo87 commented on package.json:
```diff
@@ -53,6 +53,8 @@
     "testdouble": "^1.6.0"
   },
   "dependencies": {
+    "broccoli": "^0.16.9",
```
you might want to use https://github.com/ember-cli/broccoli-builder here instead

Turbo87 commented on lib/tasks/lint.js:
```diff
@@ -0,0 +1,41 @@
+
+  var tree = new Funnel(this.project.root, {
+    include: [
+      'app/**/*.js',
```
what about `addon/**/*.js`?

Turbo87 commented on lib/tasks/lint.js:
```diff
@@ -0,0 +1,41 @@
+  var eslintTree = eslint(tree, {
+    console: {
+      log: function(message) {
+        hadError = true;
```
should `console.log()` trigger a non-zero exit code? e.g. when warnings are printed the lint command shouldn't fail, right?

alexlafroscia replied:
Hmm, I guess that would depend on the user's configuration for fail-on-warn.  What would be the best way to look that up?

BlueRaja commented:
Why was this never merged?

alexlafroscia commented:
@BlueRaja Not included because I never got an answer about whether or not to exit with a non-zero exit code when there are warnings but not errors.  See [this comment](https://github.com/ember-cli/ember-cli-eslint/pull/93#r77775500).

`lint:fix` is blocked behind [this PR](https://github.com/ember-cli/broccoli-lint-eslint/pull/49), which I need to get back to.  It's been months since I originally submitted it, so it needs to be rebased and have some of the feedback reviewed.

Turbo87 commented:
> I never got an answer about whether or not to exit with a non-zero exit code when there are warnings but not errors.

IMHO we should mimic what the original CLI tool does. if that has a non-zero exit status for warnings then we should have it too.

alexlafroscia commented:
I thought fail-on-warn was configurable but maybe not.  I'll remove it from here.

alexlafroscia commented:
Also rebased from `master`

BrianSipple commented:
@alexlafroscia `broccoli-lint-eslint` should allow for either `throwOnWarn` or `throwOnError` to be set after [this](https://github.com/ember-cli/broccoli-lint-eslint/pull/60)

alexlafroscia commented:
Will those settings throw immediately? Or after the linting is complete?  I'm curious if the linting will finish (and print out all the errors) and _then_ throw, or fail right away and not print everything.

peabnuts123 commented:
I would like to resurrect this thread as I think this is an invaluable feature that seems to be all but completed. I frequently run `ember s` purely to check linting.

Turbo87 commented:
@peabnuts123 FWIW I tend to prefer a `lint` script in `package.json` instead (e.g. https://github.com/simplabs/ember-test-selectors/blob/master/package.json#L17). Unfortunately yarn only installs launchers for direct dependencies so you might have to add ESLint as a direct dev dep too. This solution still seems massively simpler than recreating their CLI launcher in this addon for no good reason. The same is also the case for the `--fix` option, which IMHO seems out of scope for this addon.

alexlafroscia commented:
I'm personally not interested in getting this over the finish line. I agree with @Turbo87 here; having an NPM script just makes more sense. We have something like `yarn lint` and `yarn lint:fix` that will run ESLint/Prettier/whatever. 

If you're interested in picking up the torch on the PR though, @peabnuts123, feel free! However, at this point, I would close this PR because I don't think I'll be coming back to it.

peabnuts123 commented:
The reason I ended up on this PR at all was because I was under the impression that running `ember s` produced a different lint output than just `eslint`, but it sounds like I should re-confirm this behaviour. I thought `ember s` was running eslint with a specific set of flags, and assumed it would make more sense to just extract out this behaviour into an ember command instead of replicating it as an NPM script in every project. I will get back to you

peabnuts123 commented:
Okay, I can confirm that in my current project simply running: 
```bash
eslint -c ./.eslintrc.js ./
```
blocks the terminal, and produces no output. It seems like eslint expects you to specify some folders or files in particular, meaning that, in order to recreate the functionality of `ember s`, one would have to specify a command such as:

```bash
eslint -c ./.eslintrc.js app/ tests/ mirage/
``` 

However, even this does not produce the same output as `ember s`, which pretty much confirms my assumption that `ember s` is running a specific configuration of eslint, as per my previous comment, and this would be nontrivial to replicate in `package.json` and even less-so to copy it across multiple projects / update it if things ever changes, etc.

Turbo87 commented:
@peabnuts123 `ember s` is only running ESLint on the files in your app, but not on e.g. `ember-cli-build.js` itself. as you can see in https://github.com/simplabs/ember-test-selectors/blob/master/package.json#L17 we are also manually listing all the folders and files we want to have linted. if you just use `./` you will potentially also lint everything inside `tmp` and `node_modules` which is certainly not what you want. 

IMHO having something like `ember lint` in this addon will increase the complexity of this addon even further without any real benefits. since @alexlafroscia seems to agree with me I'll close his PR now. we could however add something about this to the README in a separate PR.

alvarocastro commented:
I'm having the same issue as @peabnuts123, the linter errors found when using `ember s` are different than when using `eslint <file>`. I tried using the `-c` param for eslint but with no success.

`ember s` seems to find a lot more errors than doing `eslint` and can't figure out why is it.

An example is this piece of code:
```
import {
  Model,
  attr,
  key,
  hasMany
} from 'ember-orbit';
```

The indentation there are 2 spaces, `ember s` output for that file is:
```
  2:1  error  Expected indentation of 1 tab but found 2 spaces  indent
  3:1  error  Expected indentation of 1 tab but found 2 spaces  indent
  4:1  error  Expected indentation of 1 tab but found 2 spaces  indent
  5:1  error  Expected indentation of 1 tab but found 2 spaces  indent
```

But when running eslint for that single file it throws no errors (just for science removed the semicolon and it does throw an error for that, so the eslint is somewhat working)

My rule for indent is as follows:
```
"indent": [2, "tab", {"SwitchCase": 1, "VariableDeclarator": 1}],
```

peabnuts123 commented:
As an indentation of 1 tab is not a default setting in eslint but missing-semicolons is, it seems like your `eslint` command is not reading your .eslintrc config file. You will need to use the `-c` parameter in order for this to work, so make sure that is working / changing the results. You should `cd` into the directory containing your .eslintrc file and then specify the folder from there: 

```
$ cd ~/Documents/my-ember-app
$ eslint -c ./.eslintrc app/path/to/file.js
```

Turbo87 commented:
@alvarocastro make sure that you are using the same version of `eslint` that `ember-cli-eslint` is using internally. If the rule linting code was changed internally then different versions can result in different errors.
